### PR TITLE
Collinearity Analysis among #func,#unit,#proptery_based,#fuzz_target

### PR DESCRIPTION
**Pull request by JeremyJC67**
Add collinearity analysis part of 4 predictors in our regression model

### CODE REVIEW

File: data/meta/oss_fuzz_python.jsonl [hunk 1]
```diff
@@ -1,264 +1,264 @@
-{"homepage": "https://github.com/hynek/pem", "language": "python", "main_repo": "https://github.com/hynek/pem", "fuzzing_engines": ["libfuzzer"], "sanitizers": ["address"], "vendor_ccs": ["david@adalogics.com", "adam@adalogics.com", "arthur.chan@adalogics.com"], "#fuzz_target": 1, "repo_id": "hynek/pem"}
+{"fuzzing_engines": ["libfuzzer"], "homepage": "https://github.com/abseil/abseil-py", "language": "python", "main_repo": "https://github.com/abseil/abseil-py", "sanitizers": ["address", "undefined"], "vendor_ccs": ["david@adalogics.com"], "#fuzz_target": 1, "repo_id": "abseil/abseil-py"}
```
Comment (EYH0602) [inline data/meta/oss_fuzz_python.jsonl]: please don't modify this file, suggest do `git restore data/meta/oss_fuzz_python.jsonl`

File: Analysis/collinearity_analysis.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import pandas as pd
```
Comment (EYH0602) [inline Analysis/collinearity_analysis.py]: format the file

File: Analysis/collinearity_analysis.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+from scipy.stats import pearsonr
+import matplotlib.pyplot as plt
+import seaborn as sns
+import fire
+def main():
+    dataset_path = './data/static_analysis_data/dataset.csv'
```
Comment (EYH0602) [inline Analysis/collinearity_analysis.py]: make this an input parameter

File: Analysis/collinearity_analysis.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    predictors = ["unit test functions number","property-based test functions number","fuzz target number"]
+    
+    print(f"Correlation Coefficient of funcs number and unit test functions number : {correlation_unit_funcs:.4f}")
+    print(f"Correlation Coefficient of funcs number and property-based test functions number : {correlation_proptery_based_funcs:.4f}")
+    print(f"Correlation Coefficient of funcs number and fuzz target number : {correlation_fuzz_target_funcs:.4f}")
+
+    plt.figure(figsize=(10, 6))
+    plt.bar(predictors, correlations, color=['blue', 'green', 'red'])
+
+    plt.title('Correlation with Funcs Number')
```
Comment (EYH0602) [inline Analysis/collinearity_analysis.py]: Do not include a title inside the figure; instead, the caption should
serve this function.

File: Analysis/collinearity_analysis.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    
+    columns_of_interest = ['#funcs', '#unit', '#proptery_based', '#fuzz_target']
+
+    correlation_matrix = dataset[columns_of_interest].corr()
+
+    sns.set_theme(style="whitegrid")
+
+    plt.figure(figsize=(10, 8))
+    sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', linewidths=.5)
+    plt.title('Predictors Correlation Heatmap')
```
Comment (EYH0602) [inline Analysis/collinearity_analysis.py]: remove title

**Review by EYH0602 — CHANGES_REQUESTED** [review]

**Comment by EYH0602** [comment]
don't commit the `.png` files, please remove

File: Analysis/collinearity_analysis.py (hunk 2 above)
Comment (EYH0602) [reply Analysis/collinearity_analysis.py]: also, please just put this file in `src/`

**Review by EYH0602 — CHANGES_REQUESTED** [review]

**Comment by JeremyJC67** [comment]
OK

Yifeng He ***@***.***> 于2024年2月21日周三 19:23写道：

File: Analysis/collinearity_analysis.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    plt.bar(predictors, correlations, color=['blue', 'green', 'red'])
+
+    plt.title('Correlation with Funcs Number')
+    plt.xlabel('Predictors')
+    plt.ylabel('Pearson Correlation Coefficient')
+
+    for i, corr in enumerate(correlations):
+        plt.text(i, corr if corr > 0 else corr - 0.02, f'{corr:.4f}', ha = 'center', va = 'bottom' if corr > 0 else 'top')
+
+    plt.savefig("./Analysis/correlation_coefficient.png",dpi=500)
```
Comment (EYH0602) [inline Analysis/collinearity_analysis.py]: use input parameter for output file names too

File: Analysis/collinearity_analysis.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    proptery_based_test_funcs_number = dataset['#proptery_based']
+    fuzz_target_number = dataset['#fuzz_target']
+
+
+    correlation_unit_funcs = pearsonr(funcs_number, unit_test_funcs_number)[0]
+    correlation_proptery_based_funcs = pearsonr(funcs_number, proptery_based_test_funcs_number)[0]
+    correlation_fuzz_target_funcs = pearsonr(funcs_number, fuzz_target_number)[0]
+
+    correlations = [correlation_unit_funcs,correlation_proptery_based_funcs,correlation_fuzz_target_funcs]
+    predictors = ["unit test functions number","property-based test functions number","fuzz target number"]
```
Comment (EYH0602) [inline Analysis/collinearity_analysis.py]: names too long, making x-axis on the figure hard to read

**Review by EYH0602 — CHANGES_REQUESTED** [review]

File: Analysis/collinearity_analysis.py (hunk 7 above)
Comment (JeremyJC67) [reply Analysis/collinearity_analysis.py]: OK， I will change it

> names too long, making x-axis on the figure hard to read

File: src/collinearity_analysis.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+This script make collinearity analysis based on 4 predictors for our linear regression
+"""
+
+import pandas as pd
+from scipy.stats import pearsonr
+import matplotlib.pyplot as plt
+import seaborn as sns
+import fire
+
+def collinearity_analysis(dataset_path:str,barchart_name:str,heatmap_name:str):
```
Comment (EYH0602) [inline src/collinearity_analysis.py]: please format using [black formatter](https://marketplace.visualstudio.com/items?itemName=ms-python.black-formatter). You can do this by installing this vscode extension and right click then choosing "format document".

**Review by EYH0602 — CHANGES_REQUESTED** [review]

**Comment by JeremyJC67** [comment]
OK

File: src/collinearity_analysis.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    
+def main():
+    dataset_path = './data/static_analysis_data/dataset.csv'
+    barchart_name = "correlation_coefficient"
+    heatmap_name = "collinearity_heatmap"
+    collinearity_analysis(dataset_path,barchart_name,heatmap_name)
+    
+    
+if __name__ == "__main__":
+    fire.Fire(main)
```
Comment (EYH0602) [inline src/collinearity_analysis.py]: `fire.Fire` is for automatically generating command line interfaces (CLIs) from absolutely any Python function/object. 
Suggest change:

```py
def collinearity_analysis(
    dataset_path: str = "./data/static_analysis_data/dataset.csv",
    barchart_name: str = "correlation_coefficient",
    heatmap_name: str = "collinearity_heatmap",
):
    ...

if __name__ == "__main__":
    fire.Fire(collinearity_analysis)
```

**Review by EYH0602 — CHANGES_REQUESTED** [review]

**Comment by JeremyJC67** [comment]
OK I agree!

File: src/collinearity_analysis.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    plt.savefig(f"./src/{barchart_name}.png",dpi=500)
+    plt.show()
+    
+    columns_of_interest = ['#funcs', '#unit', '#proptery_based', '#fuzz_target']
+    correlation_matrix = dataset[columns_of_interest].corr()
+
+    sns.set_theme(style="whitegrid")
+    plt.figure(figsize=(10, 8))
+    sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', linewidths=.5)
+    plt.savefig(f"./src/{heatmap_name}.png",dpi=500)
```
Comment (EYH0602) [inline src/collinearity_analysis.py]: don't save images to `src/` which stands for "source code", I suggest making output directory also a cli input

Comment (EYH0602) [inline src/collinearity_analysis.py]: save figures to PDF format and reduce the white surrounding white space, use `format="pdf", bbox_inches="tight"`

**Review by EYH0602 — CHANGES_REQUESTED** [review]

**Comment by JeremyJC67** [comment]
Ok, I will do it.

File: src/collinearity_analysis.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import seaborn as sns
+import fire
+
+def collinearity_analysis(dataset_path:str,barchart_name:str,heatmap_name:str):
+    
+    dataset = pd.read_csv(dataset_path)
+
+    funcs_number = dataset['#funcs']
+    unit_test_funcs_number = dataset['#unit']
+    proptery_based_test_funcs_number = dataset['#proptery_based']
```
Comment (EYH0602) [inline src/collinearity_analysis.py]: proptery -> property, please also fix this in `src/static.py` line 163.

**Review by EYH0602 — CHANGES_REQUESTED** [review]

File: src/collinearity_analysis.py (hunk 8 above)
Comment (JeremyJC67) [reply src/collinearity_analysis.py]: really helpful,thx!

**Review by EYH0602 — APPROVED** [review]

File: src/collinearity_analysis.py (hunk 11 above)
Comment (EYH0602) [reply src/collinearity_analysis.py]: make sure other cases in the script are also fixed

**Review by acharyadarshan — APPROVED** [review]

**Pull request merged**